docs: use scale testing utility (#12643)
coder/coder: docs/admin/scale.md

support, you may wish to use our internal scale testing utility against your own
environments.

> Note: This utility is experimental. It is not subject to any compatibility
> guarantees, and may cause interruptions for your users. To avoid potential
> outages and orphaned resources, we recommend running scale tests on a
> secondary "staging" environment or a dedicated

> > [Kubernetes playground cluster](https://github.com/coder/coder/tree/main/scaletest/templates).

Reviewer: That template is for the scaletest runner -- if you want to spin up a k8s playground you instead want to look in the `terraform` dir.

```suggestion
> [Kubernetes playground cluster](https://github.com/coder/coder/tree/main/scaletest/terraform).
```

		--timeout 5h \
		--job-timeout 5h \
		--no-cleanup \
		--output json:"${SCALETEST_RESULTS_DIR}/create-workspaces.json"

# Run `coder exp scaletest create-workspaces --help` for all usage
```

The command does the following:

1. Create `${SCALETEST_PARAM_NUM_WORKSPACES}` workspaces concurrently
   (concurrency level: `${SCALETEST_PARAM_CREATE_CONCURRENCY}`) using the
   template `${SCALETEST_PARAM_TEMPLATE}`.
1. Leave workspaces running to use in next steps (`--no-cleanup` option).
1. Store provisioning results in JSON format.
1. If you don't want the creation process to be interrupted by any errors, use
   the `--retry 5` flag.

### Traffic Generation

Given an existing set of workspaces created previously with `create-workspaces`,
the following command will generate traffic similar to that of Coder's Web
Terminal against those workspaces.

```shell
# Produce load at about 1000MB/s (25MB/40ms).
coder exp scaletest workspace-traffic \
	--template "${SCALETEST_PARAM_GREEDY_AGENT_TEMPLATE}" \
	--bytes-per-tick $((1024 * 1024 * 25)) \
	--tick-interval 40ms \
	--timeout "$((delay))s" \
	--job-timeout "$((delay))s" \
	--scaletest-prometheus-address 0.0.0.0:21113 \

>   --target-workspaces "0:100" \

Reviewer: ```suggestion
    --target-workspaces "0:100" \
```

	--trace=false \
  --output json:"${SCALETEST_RESULTS_DIR}/traffic-${type}-greedy-agent.json"
```

Traffic generation can be parametrized:

1. Send `bytes-per-tick` every `tick-interval`.
1. Enable tracing for performance debugging.
1. Target a range of workspaces with `--target-workspaces 0:100`.
1. For dashboard traffic: Target a range of users with `--target-users 0:100`.

> 1. Store provisioning results in JSON format.

Reviewer: ```suggestion
1. Store provisioning results in JSON format.
1. Expose a dedicated Prometheus address (`--scaletest-prometheus-address`) for scaletest-specific metrics.
```

The `workspace-traffic` supports also other modes - SSH traffic, workspace app:

1. For SSH traffic: Use `--ssh` flag to generate SSH traffic instead of Web
   Terminal.
1. For workspace app traffic: Use `--app [wsdi|wsec|wsra]` flag to select app
   behavior. (modes: _WebSocket discard_, _WebSocket echo_, _WebSocket read_).

### Cleanup

The scaletest utility will attempt to clean up all workspaces it creates. If you
wish to clean up all workspaces, you can run the following command:

```shell
coder exp scaletest cleanup \
	--cleanup-job-timeout 2h \
	--cleanup-timeout 15min
```

This will delete all workspaces and users with the prefix `scaletest-`.

## Scale testing template

Consider using a dedicated
[scaletest-runner](https://github.com/coder/coder/tree/main/scaletest/templates/scaletest-runner)
template alongside the CLI utility for testing large-scale Kubernetes clusters.

The template deploys a main workspace with scripts used to orchestrate Coder,
creating workspaces, generating workspace traffic, or load-testing workspace
apps.

### Parameters

The _scaletest-runner_ offers the following configuration options:

> - Workspace template selecting Kubernetes cluster size:
>   minimal/small/medium/large (_default_: minimal)

Reviewer: ```suggestion
- Workspace size selection:
  minimal/small/medium/large (_default_: minimal, which contains just enough resources for a Coder agent to run without additional workloads)
```

- Number of workspaces
- Wait duration between scenarios or staggered approach

The template exposes parameters to control the traffic dimensions for SSH
connections, workspace apps, and dashboard tests:

- Traffic duration of the load test scenario
- Traffic percentage of targeted workspaces
- Bytes per tick and tick interval
- _For workspace apps_: modes (echo, read random data, or write and discard)

Scale testing concurrency can be controlled with the following parameters:

- Enable parallel scenarios - interleave different traffic patterns (SSH,
  workspace apps, dashboard traffic, etc.)
- Workspace creation concurrency level (_default_: 10)
- Job concurrency level - generate workspace traffic using multiple jobs
  (_default_: 0)
- Cleanup concurrency level

### Kubernetes cluster

Depending on the traffic projections, operators can deploy different sample

> clusters to perform scale tests. It is recommended to learn how to operate the

Reviewer: The workspace template doesn't control the cluster size; this is up to operators to control.

Author: Rephrased the paragraph 👍

scaletest-runner before running it against the staging cluster (or production at
your own risk).

There are a few cluster options
[available](https://github.com/coder/coder/tree/main/scaletest/templates/scaletest-runner):

> | Cluster size | vCPU | Memory | Persisted storage | Details                                               |

Reviewer: ```suggestion
| Workspace size | vCPU | Memory | Persisted storage | Details                                               |
```

| small        | 1    | 1 Gi   | None              |                                                       |
| medium       | 2    | 2 Gi   | None              | Medium-sized cluster offers the greedy agent variant. |
| large        | 4    | 4 Gi   | None              |                                                       |

#### Greedy agent

The greedy agent variant is a template modification that makes the Coder agent
transmit large metadata (size: 4K) while reporting stats. The transmission of
large chunks puts extra overhead on coderd instances and agents when handling
and storing the data.

Use this template variant to verify limits of the cluster performance.

### Observability

During scale tests, operators can monitor progress using a Grafana dashboard.
Coder offers a comprehensive overview
[dashboard](https://github.com/coder/coder/blob/main/scaletest/scaletest_dashboard.json)
that can seamlessly integrate into the internal Grafana deployment.

This dashboard provides insights into various aspects, including:

- Utilization of resources within the Coder control plane (CPU, memory, pods)

> - Database performance metrics (CPU, memory, I/O, connections, queries)

Reviewer: Note: the coder deployment needs to have `CODER_PROMETHEUS_COLLECT_DB_METRICS=true`

Author: Noted
